File: problems/fannkuch/fannkuch.c
// fannkuch-redux benchmark
// Derived from the Computer Language Benchmarks Game
// https://benchmarksgame-team.pages.debian.net/benchmarksgame/

#include <stdio.h>
#include <stdlib.h>

int fannkuch(int n) {
  int *perm = malloc(n * sizeof(int));
  int *perm1 = malloc(n * sizeof(int));
  int *count = malloc(n * sizeof(int));

  int max_flips = 0;
  int checksum = 0;
  int perm_count = 0;

  // Initialize
  for (int i = 0; i < n; i++) {
    perm1[i] = i;
  }

  int r = n;

  while (1) {
    // Copy perm1 to perm
    while (r != 1) {
      count[r - 1] = r;
      r--;
    }

    for (int i = 0; i < n; i++) {
      perm[i] = perm1[i];
    }

    // Count flips
    int flips = 0;
    int k;
    while ((k = perm[0]) != 0) {
      // Reverse first k+1 elements
      int i = 0;
      int j = k;
      while (i < j) {
        int temp = perm[i];
        perm[i] = perm[j];
        perm[j] = temp;
        i++;
        j--;
      }
      flips++;
    }

    if (flips > max_flips) {
      max_flips = flips;
    }

    // Checksum
    if (perm_count % 2 == 0) {
      checksum += flips;
    } else {
      checksum -= flips;
    }
    perm_count++;

    // Generate next permutation
    while (1) {
      if (r == n) {
        free(perm);
        free(perm1);
        free(count);
        // Return combined result for verification
        return checksum * 1000 + max_flips;
      }

      int perm0 = perm1[0];
      for (int i = 0; i < r; i++) {
        perm1[i] = perm1[i + 1];
      }
      perm1[r] = perm0;

      count[r]--;
      if (count[r] > 0) {
        break;
      }
      r++;
    }
  }
}

int main(int argc, char **argv) {
  // Use argc to prevent constant folding, default n=10
  int n = 10 + (argc - 1);

  int result = fannkuch(n);
  int checksum = result / 1000;
  int max_flips = result % 1000;

  // Expected for n=10: checksum=73196, max_flips=38
  // With argc=1, n=10
  if (argc == 1) {
    if (checksum == 73196 && max_flips == 38) {
      return 0;
    }
  }

  // For other n values, just verify it ran
  if (max_flips > 0) {
    return 0;
  }
  return 1;
}

File: problems/fannkuch/fannkuch.rs
// fannkuch-redux benchmark
// Derived from the Computer Language Benchmarks Game

fn fannkuch(n: usize) -> (i32, i32) {
    let mut perm = vec![0usize; n];
    let mut perm1: Vec<usize> = (0..n).collect();
    let mut count = vec![0usize; n];
    
    let mut max_flips = 0i32;
    let mut checksum = 0i32;
    let mut perm_count = 0i32;
    
    let mut r = n;
    
    loop {
        // Copy perm1 to perm
        while r != 1 {
            count[r - 1] = r;
            r -= 1;
        }
        
        perm.copy_from_slice(&perm1);
        
        // Count flips
        let mut flips = 0i32;
        loop {
            let k = perm[0];
            if k == 0 {
                break;
            }
            // Reverse first k+1 elements
            let mut i = 0;
            let mut j = k;
            while i < j {
                perm.swap(i, j);
                i += 1;
                j -= 1;
            }
            flips += 1;
        }
        
        if flips > max_flips {
            max_flips = flips;
        }
        
        // Checksum
        if perm_count % 2 == 0 {
            checksum += flips;
        } else {
            checksum -= flips;
        }
        perm_count += 1;
        
        // Generate next permutation
        loop {
            if r == n {
                return (checksum, max_flips);
            }
            
            let perm0 = perm1[0];
            for i in 0..r {
                perm1[i] = perm1[i + 1];
            }
            perm1[r] = perm0;
            
            count[r] -= 1;
            if count[r] > 0 {
                break;
            }
            r += 1;
        }
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    // Use argc to prevent constant folding, default n=10
    let n = 10 + (args.len() - 1);
    
    let (checksum, max_flips) = fannkuch(n);
    
    // Expected for n=10: checksum=73196, max_flips=38
    if args.len() == 1 {
        if checksum == 73196 && max_flips == 38 {
            std::process::exit(0);
        }
    }
    
    // For other n values, just verify it ran
    if max_flips > 0 {
        std::process::exit(0);
    }
    std::process::exit(1);
}
